get_routers/get_routers.py

```diff
@@ -0,0 +1,83 @@
+"""
+The following Python code sample shows how to access the
+"routers" endpoint of the NCM REST API using the
+Python "requests" module with paging. It makes use of NCM
+and CP credential headers to authenticate and access the
+router data. The router_id is the key and the router_group,
+router_name, and router_state are the values.
+"""
+
+import json
+import requests
+from datetime import datetime, timezone, timedelta
+
+headers = {
+    'X-CP-API-ID': '...',
+    'X-CP-API-KEY': '...',
+    'X-ECM-API-ID': '...',
+    'X-ECM-API-KEY': '...',
+    'Accept': '*/*',
+    'Content-Type': 'application/json',
+}
+
+
+account = ''  # your ncm account id here.
+base_url = 'https://www.cradlepointecm.com'
+
+
+def get_time():
+    """ Routers updated within the last 10 minutes. """
+    return (datetime.now(timezone.utc).astimezone()
+            - timedelta(minutes=10)).isoformat()
+
+
+def get_router_info(time):
+    """ Get router information """
+
+    url = base_url + f'/api/v2/routers/?limit=500&account={account}'
+
+    # we ultimately want to filter on state_updated_at to reduce api calls.
```

Comment:
Why isn't this condition (which should be a >, not =) implemented in the script? get_time() is never called.

Reply:
We are currently not allowed to filter on state_updated_at so I provided a commented out example for the future.

url = base_url + f'/api/v2/routers/?limit=500&account={account}\
&state_updated_at>{last_10_minutes}'

{'exception': {'type': 'error', 'message': '&#39;exact&#39; is not an allowed filter on the &#39;state_updated_at&#39; field.'}, 'errors': [{'path': '/api/v2/routers/'}]}

get_time() is set below as a variable named last_10_minutes (it was named time before but i just renamed it...) that is passed into the get_router_info function.

def get_router_info(last_10_minutes):
    """ Get router information """

if __name__ == '__main__':
    last_10_minutes = get_time()
    routers = get_router_info(last_10_minutes)
    results = dictionary_to_json(routers)
    print(results)

Reply:
According to the documentation, filtering on state_updated_at is allowed.  Here's a real-life call that succeeded today: GET /api/v2/routers/?limit=1000&state_updated_at__gt=2019-04-17+19%3A23%3A20.381951&state_updated_at__lt=2019-04-17+19%3A24%3A20.708449

Reply:
Thank you!.. That is the piece that I was missing...

account = '' # your ncm account id here.
base_url = 'https://www.cradlepointecm.com'
updated_in = 0 # Routers updated within the last 0 (now) minutes.


def get_time():
    """ Returns the correct time format minus the updated_in minutes. """
    return (datetime.now(timezone.utc).astimezone()
            - timedelta(minutes=updated_in)).isoformat()


def get_router_info(now):
    """ Get router information """

    url = base_url + f'/api/v2/routers/?limit=500&account={account}\
    &state_updated_at__lt={now}'

    routers = {}